src/ser.rs

```diff
@@ -401,115 +209,113 @@ where
         self.serialize_tuple(len)
     }
 
     #[inline]
     fn serialize_tuple_variant(
         self,
-        name: &'static str,
-        variant_index: u32,
+        _name: &'static str,
+        _variant_index: u32,
         variant: &'static str,
         len: usize,
-    ) -> Result<&'a mut Serializer<W>> {
-        if self.enum_as_map {
-            self.write_u64(5, 1u64)?;
-            variant.serialize(&mut *self)?;
-            self.serialize_tuple(len)
-        } else {
-            self.write_u64(4, (len + 1) as u64)?;
-            self.serialize_unit_variant(name, variant_index, variant)?;
-            Ok(self)
-        }
+    ) -> Result<Self::SerializeTupleVariant, Self::Error> {
+        enc::MapStartBounded(1).encode(&mut self.writer)?;
+        variant.encode(&mut self.writer)?;
+        enc::ArrayStartBounded(len).encode(&mut self.writer)?;
+        Ok(BoundedCollect { ser: self })
     }
 
     #[inline]
-    fn serialize_map(self, len: Option<usize>) -> Result<CollectionSerializer<'a, W>> {
-        self.serialize_collection(5, len)
-    }
-
-    #[cfg(not(feature = "std"))]
-    fn collect_str<T: ?Sized>(self, value: &T) -> Result<()>
-    where
-        T: core::fmt::Display,
-    {
-        use crate::write::FmtWrite;
-        use core::fmt::Write;
-
-        let mut w = FmtWrite::new(&mut self.writer);
-        write!(w, "{}", value)?;
-        Ok(())
+    fn serialize_map(self, len: Option<usize>) -> Result<Self::SerializeMap, Self::Error> {
+        if let Some(len) = len {
+            enc::MapStartBounded(len).encode(&mut self.writer)?;
+        } else {
+            enc::MapStartUnbounded.encode(&mut self.writer)?;
+        }
+        Ok(Collect {
+            bounded: len.is_some(),
+            ser: self,
+        })
     }
 
     #[inline]
-    fn serialize_struct(self, _name: &'static str, len: usize) -> Result<StructSerializer<'a, W>> {
-        self.write_u64(5, len as u64)?;
-        Ok(StructSerializer { ser: self, idx: 0 })
+    fn serialize_struct(
+        self,
+        _name: &'static str,
+        len: usize,
+    ) -> Result<Self::SerializeStruct, Self::Error> {
+        enc::MapStartBounded(len).encode(&mut self.writer)?;
+        Ok(BoundedCollect { ser: self })
     }
 
     #[inline]
     fn serialize_struct_variant(
         self,
-        name: &'static str,
-        variant_index: u32,
+        _name: &'static str,
+        _variant_index: u32,
         variant: &'static str,
         len: usize,
-    ) -> Result<StructSerializer<'a, W>> {
-        if self.enum_as_map {
-            self.write_u64(5, 1u64)?;
-        } else {
-            self.writer.write_all(&[4 << 5 | 2]).map_err(|e| e.into())?;
-        }
-        self.serialize_unit_variant(name, variant_index, variant)?;
-        self.serialize_struct(name, len)
+    ) -> Result<Self::SerializeStructVariant, Self::Error> {
+        enc::MapStartBounded(1).encode(&mut self.writer)?;
```

Comment:
why are you creating two maps?

Reply:
`MyStruct::Pointer { pos: 4, len: 2 }` becomes:

```json
{
  "Pointer": {
    "pos": 4,
    "len": 2
  }
}
```

Reply:
I see